File: projects/03-realtime-pubsub/src/protocol.rs
//! The wire protocol and connection identity — the glue types every module
//! speaks. This is scaffolding, not a vertical challenge: the message *shapes*
//! are given so the WebSocket handler, the hub, and the cluster bridge all agree
//! on a vocabulary. Extend it (e.g. add `ping`/`pong` app-level frames, an ack,
//! or a `history` request) as the SPEC's protocol checklist asks.

use std::sync::atomic::{AtomicU64, Ordering};

use serde::{Deserialize, Serialize};

pub type Topic = String;

/// A process-unique id for one open WebSocket connection. Cheap to copy; used as
/// the key a subscriber is tracked under in the hub and presence registry.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct ConnId(u64);

impl ConnId {
    /// Mint the next id. Monotonic for the lifetime of the process.
    pub fn next() -> Self {
        static COUNTER: AtomicU64 = AtomicU64::new(1);
        ConnId(COUNTER.fetch_add(1, Ordering::Relaxed))
    }

    pub fn as_u64(self) -> u64 {
        self.0
    }
}

impl std::fmt::Display for ConnId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "conn-{}", self.0)
    }
}

/// Messages a client sends *to* the server (deserialized from each text frame).
///
/// The `type` field is the tag, e.g. `{"type":"subscribe","topic":"room1"}`.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    /// Join a topic; future publishes to it will be delivered to this socket.
    Subscribe { topic: String },
    /// Leave a topic.
    Unsubscribe { topic: String },
    /// Broadcast `payload` to everyone subscribed to `topic`.
    Publish {
        topic: String,
        payload: serde_json::Value,
    },
    /// Application-level liveness ping (`{"type":"heartbeat"}`). Browsers can't
    /// send WebSocket protocol ping frames from JS, so a live-but-idle client
    /// sends this on a timer to prove it's still here and refresh its presence
    /// TTL. Carries no payload — its arrival *is* the signal.
    Heartbeat,
}

/// Messages the server sends *to* a client (serialized into a text frame).
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    /// A broadcast delivered on a topic this client subscribes to.
    Message {
        topic: String,
        payload: serde_json::Value,
    },
    /// The current membership of a topic (V3).
    Presence { topic: String, members: Vec<String> },
    /// A protocol-level error (bad frame, unknown topic, over a limit, …).
    Error { reason: String },
}

File: projects/03-realtime-pubsub/src/realtime_pubsub/protocol.py
"""The wire protocol and connection identity — the glue types every module
speaks.

This is scaffolding, not a vertical challenge: the message *shapes* are given so
the WebSocket handler, the hub, and the cluster bridge all agree on a
vocabulary. Extend it (an ack, a `history` request, app-level ping/pong) as the
SPEC's protocol checklist asks.

Python shape note: the Rust side used serde's internally-tagged enums. The
equivalent here is a **pydantic discriminated union** on the `type` field. It
buys the same two things — one `model_validate` call turns an arbitrary frame
into exactly one of the known variants, and an unknown or malformed `type` is a
`ValidationError` rather than a silently half-parsed object — plus the error
message names the offending field, which is what `routes.py` echoes back to the
client as a protocol `error`.
"""

from __future__ import annotations

from itertools import count
from typing import Annotated, Any, Literal, NewType

from pydantic import BaseModel, Field, TypeAdapter

__all__ = [
    "BroadcastMessage",
    "ClientMessage",
    "ConnId",
    "ErrorMessage",
    "Heartbeat",
    "PresenceMessage",
    "Publish",
    "ServerMessage",
    "Subscribe",
    "Topic",
    "Unsubscribe",
    "client_message_adapter",
    "conn_label",
    "next_conn_id",
]

type Topic = str

# A process-unique id for one open WebSocket connection. Used as the key a
# subscriber is tracked under in the hub and the presence registry.
#
# `NewType` over a plain `int` rather than a wrapper class: it is free at
# runtime, it stays hashable and comparable for use as a dict key, and pyright
# still refuses to let you pass a bare `int` where a ConnId belongs. The Rust
# side needed an `AtomicU64` behind this; here a module-level `itertools.count`
# is enough, because ids are only ever minted from the event loop thread and
# `next()` on a count object is atomic under the GIL anyway.
ConnId = NewType("ConnId", int)

_counter = count(1)


def next_conn_id() -> ConnId:
    """Mint the next id. Monotonic for the lifetime of the process."""
    return ConnId(next(_counter))


def conn_label(conn: ConnId) -> str:
    """Human-readable form used in logs and as the presence fallback identity."""
    return f"conn-{conn}"


# --- Client -> server ---------------------------------------------------------


class Subscribe(BaseModel):
    """Join a topic; future publishes to it will be delivered to this socket."""

    type: Literal["subscribe"]
    topic: str


class Unsubscribe(BaseModel):
    """Leave a topic."""

    type: Literal["unsubscribe"]
    topic: str


class Publish(BaseModel):
    """Broadcast `payload` to everyone subscribed to `topic`."""

    type: Literal["publish"]
    topic: str
    payload: Any


class Heartbeat(BaseModel):
    """Application-level liveness ping (`{"type":"heartbeat"}`).

    Browsers cannot send WebSocket protocol ping frames from JS, so a
    live-but-idle client sends this on a timer to prove it is still here and
    refresh its presence TTL. Carries no payload — its arrival *is* the signal.
    """

    type: Literal["heartbeat"]


type ClientMessage = Annotated[
    Subscribe | Unsubscribe | Publish | Heartbeat,
    Field(discriminator="type"),
]

# Built once at import: a TypeAdapter compiles the union's validator, and
# rebuilding it per frame would put that cost on the hot path.
client_message_adapter: TypeAdapter[ClientMessage] = TypeAdapter(ClientMessage)


# --- Server -> client ---------------------------------------------------------


class PresenceMessage(BaseModel):
    """The current membership of a topic (V3)."""

    type: Literal["presence"] = "presence"
    topic: str
    members: list[str]


class BroadcastMessage(BaseModel):
    """A broadcast delivered on a topic this client subscribes to."""

    type: Literal["message"] = "message"
    topic: str
    payload: Any


class ErrorMessage(BaseModel):
    """A protocol-level error (bad frame, over a limit, ...)."""

    type: Literal["error"] = "error"
    reason: str


type ServerMessage = BroadcastMessage | PresenceMessage | ErrorMessage
